### PR TITLE
D-Adaptation and Prodigy contrib implementations

### PR DESCRIPTION
**Pull request by adefazio**
Implementations of D-Adaptation AdamW and the related method Prodigy based on the official PyTorch implementations. I have verified that they give the same outputs as the PyTorch version on an example problem. Unit tests similar to those used on other optimizers in contrib.

https://github.com/facebookresearch/dadaptation
https://github.com/konstmish/prodigy

These two new optimizers perform learning rate adaptation, similar to Mechanic and COCOB, two optimizers already included in contrib, but by a different mechanism, and so I think these are relevant to Optax and interesting to the community. D-Adaptation won an ICML outstanding paper award and is already gaining a lot of traction in the ML community, particularly for fine-tuning diffusion models with the Prodigy variant.

### CODE REVIEW

**Comment by fabianp** [comment]
Thanks @adefazio  for the contribution!

**Review by vroulet — COMMENTED** [review]
Thanks for the contribution!
Minor comment: if s and d could have more explicit names, that could be helpful for a newcomer to investigate the algorithm but numerous algorithms have non-readable parameters like b1 or beta so that's fine as is too.

**Review by vroulet — COMMENTED** [review]
Looks perfect, thank you! Final request: could you squash your commits into one?

**Review by mtthss — APPROVED** [review]
